Add document on confidential transactions (#1228)
ElementsProject/elements: doc/elements-confidential-transactions.md

## Overview of Confidential Assets

Using Elements, the sender of a transaction can hide the amounts and
types of assets in a transaction’s outputs, in such a way that: 

1. Only the sender and receiver of the transaction can see the actual
amounts and types of assets.
2. A verifier can prove that all assets coming out of a transaction
went into it. 
3. The amounts and assets of the outputs may be revealed to a third
party, by the receiver or by the sender. 

This feature is called Confidential Assets. To create a confidential
assets transaction, the recipient generates a Confidential Address and
provides it to the sender. The sender uses the address to create a
Confidential Transaction, and “blinds” its outputs, obscuring the type
and quantity of assets while guaranteeing the integrity of the
transaction. The recipient may later “unblind” and spend the
transaction’s outputs at will. The unblinding process is called
“rewinding”, or “rewinding the range proof”, and requires the private
blinding key of the confidential address. Either the sender or a
receiver may also share a blinding key with a third party, enabling
them to view, but not to spend, the transaction’s outputs. 

> Confidential Assets does not conceal the transaction ids or indexes on

Reviewer: ```suggestion
Confidential Assets transactions do not conceal the transaction ids or indexes on
```

Author: fixed in amended commit

the inputs (the transaction graph is public, as it is with Bitcoin). 

A Confidential Transaction must also include an explicit (unblinded)
fee output, paid in the sidechain’s default asset (L-BTC for Liquid).

## Cryptographic Primitives

Elements' Confidential Assets capability is built on four cryptographic primitives

> implemented in the secpk25k1-zkp C++ library [1]:

Reviewer: You can actually insert footnote links in markdown.

https://github.blog/changelog/2021-09-30-footnotes-now-supported-in-markdown-fields/

Author: thank you fixed in amended commit

### Pedersen Commitments

In a confidential transaction, asset types and amounts are blinded
using Pedersen commitments and mapped to  “asset commitments” and
“value commitments”.  These commitments obscure both the asset types
and amounts, while allowing a verifier to prove that the sum of a
transaction’s inputs is equal to the sum of its outputs plus the
transaction fee, without needing to see the actual amounts.

### Range Proofs

When constructing a confidential transaction using Pedersen
commitments, it is possible to introduce an output with a negative
value, thereby causing coin inflation in other outputs. To prevent
this, a cryptographic range proof is included in each blinded

> transaction output. A range proof allows a verifier to prove that the
> output value lies within a specified non-negative range, without

Reviewer: ```suggestion
transaction output. A range proof allows a verifier to prove that an
output's value lies within a specified non-negative range, without
```

having to see the actual value.

### Surjection Proofs

When assets are blinded in a transaction, a verifier cannot see which
input assets are sent to which outputs. A Surjection proof [2] allows
a verifier to prove that an output’s asset appears in at least one

> input, without revealing the actual asset type. (The mapping from
> input assets to the output asset must be an “onto” function, or a
> “surjection”). Every blinded transaction output has a Surjection

Reviewer: ```suggestion
input, without revealing the actual asset type. In other words, the mapping from
input assets to the output asset must be an “onto” function, or a
“surjection”. Every blinded transaction output has a Surjection
```

Author: fixed in amended commit

proof.

### Elliptic Curve Diffie-Hellman (ECDH)

The receiver and the sender need a shared secret to blind and unblind
the outputs and to create the range proof. ECDH key exchange is used
to create the shared secret. When generating a Confidential Address,
the receiver creates a private blinding key and embeds the public key
in the address. The sender generates an ephemeral keypair, and with
the address’ public blinding key, uses ECDH to generate a shared
secret. The sender’s key for the ECDH key exchange is written into the

> “ConfidentialNonce” field of the transaction output.

Reviewer: nit: maybe format it like this? (not sure...)

```suggestion
`ConfidentialNonce` field of the transaction output.
```

> See the “Confidential Assets” paper [3] for more detailed information

Reviewer: italics?

```suggestion
See the _Confidential Assets_ paper [3] for more detailed information
```

Author: thanks, fixed in later commit

and references.

## Transaction Layout and Size Considerations

The layout and fields of a Confidential Transaction are described in
[elements-tx-format.md](./elements-tx-format.md).

Confidential Transactions are significantly larger than
non-confidential transactions due to the addition of the range and
surjection proofs.  Each blinded output adds about 4k bytes to the
transaction. As the proofs are stored in output witness fields, each
blinded output adds about 1k vBytes to the virtual size of the
transaction (please see https://en.bitcoin.it/wiki/Weight_units for
information about transaction size measurements).

### Range Proofs

A range proof proves that a value lies within a specific range. The
exact range is determined by the parameters of the range proof. A
tradeoff exists between the size of the range, and the size of the
proof’s serialization in the transaction output.  A larger range
requires a larger proof. In Elements, the default size of the range

> proof is 52 bits, allowing a value up to 2^52-1 satoshis to be

Reviewer: ```suggestion
proof is 52 bits, allowing for a value of up to 2^52-1 satoshis to be
```

Reviewer: Looks like this change was not accepted. No problem, just want to confirm that this was intentional.

represented. In Elements, it is possible to create more than 21

> million satoshis of an asset over multiple issuances, but it is not

Reviewer: Is this correct? The per-transaction limit is not 21 million right? Since, in bitcoin there are `21,000,000 btc x 100,000,000 sats / btc` total satoshis?

Author: thanks, fixed in later commit

Reviewer: Looks like it still reads `21 million satosahis`. Was this addressed?

> possible to send more than 21 million in any one output.  The range

Reviewer 2: Shouldn't it say 'not possible to send more than 21 million in any one unconfidential output' ? Because the code cannot see the blinded values

proof parameters are not part of consensus, and may be overridden and
adjusted using the `ct_bits` elements configuration parameter. Reducing
the number of bits will reduce the size of a transaction, and  also
reduce the maximum provable value of any output. 

An Elements range proof is a Borromean ring signature [4] over
possible values of each digit in the base 4 representation of an

> output value. Each digit requires the storage of 4+1 elliptic curve

Reviewer: The "4 + 1 points" is not immediately clear to me. I see the 4 from the base 4, but I don't see how it relates to the ecliptic curve point, or where the 1 comes from. Could you say 5, or add a bit more detail to how this calculation came about?

Author: I've written 5, with a reference to the paper.

points. Not including a fixed size header for the range proof, the
space required for a range proof in Elements is approximately 80 bytes
per bit of precision (default 52). 

The [secp256k1-zkp range proof implementation](https://github.com/BlockstreamResearch/secp256k1-zkp/blob/master/include/secp256k1_rangeproof.h)

> supports a maximum range of 64 bits, which requires a maximum of 5134 bytes. 

Reviewer: What is the 5134 bytes here referring to? Are these the vBytes in the transaction? Whereas "64 bits" is the size of the number over which the proof is being done?

Reviewer 3: They are witness bytes, not vbytes, and yes, 64 bits is the size of the number.

Author: I'm listing the actual sizes of the objects in bytes. The first paragraph says that it's part of witness fields.

Bulletproofs are another type of range proof that can be used for
confidential transactions. Bulletproofs are much smaller than the

> range proofs currently used in Elements. See [5] for more

Reviewer: Perhaps something more like:
```suggestion
range proofs currently used in Elements, and there is ongoing research being done on how
they could be added to Elements in the future. See [5] for more
```

information.

### Surjection Proofs

A surjection proof is a ring signature over the differences between
input and output commitments. The space required for a surjection
proof in Elements is proportional to the number of transaction inputs.

> Each input adds approximately 40 bytes to the surjection proof.  In

Reviewer: The size of the surjection proof mainly matters for transaction size and fee calculations, correct?

In that case, should this measurement be stated in vBytes?

Reviewer 3: I think vbytes are confusing. It's better to just say "witness bytes" and users can understand that they should divide by 4 if they think in terms of vbytes.

Author: In the size descriptions for the fields, I'm listing the actual sizes of the proofs in bytes. In the first paragraph I point out that the proofs are part of the witness fields. I've left it up to the reader to interpret the measurement of vbytes/witness bytes from the actual size and where they are.

4. `rawblindrawtransaction` - can be used to blind
a transaction whose inputs are not from the local wallet; input
blinding factors must be provided.
5. `signrawtransaction` - signs inputs using the wallet. At this
point, the transaction may be broadcasted, using `sendrawtransaction`.

In the construction of a Pedersen Commitment, random blinding factors
are used in every transaction output value commitment except the
last. The last blinding factor is chosen so that the blinding factors
(and therefore the commitments) sum to zero. As a consequence, when
blinding a transaction, there must be at least one output available to
"balance the blinding factors". The Elements RPCs
`blindrawtransaction` and `blindrawtransaction` may add an additional
zero-valued output.

## Asset Issuances and Reissuances

An asset issuance creates a non-zero amount of a new asset, and zero
or more reissuance  token that may be used to create more of the same
asset at a later time. Reissuance tokens are also called "inflation
keys".  

In Elements, there are four types of transaction inputs:

1. "typical" inputs that spend UTXOs
2. coinbase inputs
3. peg-ins
4. asset issuances/reissuances.  

> An asset issuance input defines the id of a new asset some non-zero amount

Reviewer: ```suggestion
An asset issuance input defines the ID of a new asset, some non-zero amount
```

Reviewer: It appears as though the comma was not added. Was this intentional? I feel as through there should be a comma after "asset".

Author: fixed

of the asset to be issued, and zero or more reissuance tokens. While

> the id of the asset issued must be explicit (it is a property derived

Reviewer: ```suggestion
the id of the asset must be explicit (it is a property derived
```

### Raw Transactions

`analyzepsbt`
Provides information about the blinding status of a PSET’s outputs.

`createpsbt`
All blinding parameters may be set explicitly when defining a PSET.

`decoderawtransaction`
Displays a JSON representation of a raw transaction. Useful for
inspecting confidential parameters. When the wallet is able to unblind
a parameter, the unblinded value will replace the blinded value. The
"ConfidentialNonce" parameter will be displayed as "commitmentnonce".

`gettransaction`
For a transaction in this wallet, displays properties of the
transaction, including some confidential parameters such as asset and
value blinders ("blinding factors").

`getrawtransaction <txid> true`
With `verbose=true` and `txindex=1`, `getrawtransaction` will display
public details of a transaction’s confidential parameters. Use
`unblindrawtransaction` followed by `decoderawtransaction` to unblind and
decode the transaction’s private parameters.

`blindrawtransaction`
Blinds the outputs of a raw transaction (as might be created by
`createrawtransaction`), using only wallet inputs.

`rawblindrawtransaction`
Blinds the outputs of a raw transaction (as might be created by

> `createrawtransaction`). This RPC requires all blinding factors be

Reviewer: ```suggestion
`createrawtransaction`). This RPC requires that all blinding factors be
```

provided explicitly. 

### Wallet  - Keys and Addresses

`getnewaddress`
By default, generates a confidential address encoded as blech32 (see
"Confidential Addresses" section above). The public key is embedded in
the address along with the ScriptPubKey. A confidential address is a
tuple (confidential_key, unconfidential address). 

`getaddressinfo`
Displays the (public) confidential and unconfidential properties of an address.

`dumpblindingkey`
Reveals the private blinding key for a confidential address. A
third-party will need this key to unblind transactions (see
"Third-party Unblinding" below). 

`dumpissuanceblindingkey`
Reveals the private blinding key that was used to blind the amounts on
an issuance input. This key is required when using reissuance tokens. 

`dumpmasterblindingkey`
Reveals the wallet’s master blinding key from which all blinding keys for generated addresses are derived.  See SLIP-007.

`importaddress`
A confidential address may be imported at any time. However, in order
to unblind outputs for a confidential address, it is necessary to also

> import the blinding key for that address’ confidential_key. See

Reviewer: What is a `confidential_key`? This is the first time I've seen the term mentioned.

Could this read:
```suggestion
import the blinding key for that address. See
```

Reviewer: Update: after reading the whole document, I see that it is defined later. However, I still believe this simplification makes sense.

the `importblindingkey` RPC.

`importblindingkey`
Imports the private blinding key associated with an address.

`importissuanceblindingkey`
Imports a private blinding key that may be used to unblind the amounts
on an issuance input, or reissue additional amounts of an asset (using
reissuance tokens). 

`importmasterblindingkey`
***Use with caution!*** Importing a master blinding key into a wallet will
erase the existing master blinding key, potentially making all
addresses owned by that wallet unspendable.

### Wallet - Transactions

`blindrawtransaction`
Blinds the outputs of a raw transaction (as might be created by `createrawtransaction`)

`unblindrawtransaction`
Unblinds a raw transaction, using addresses and blinding keys known to
the wallet (they must have been previously imported, see "Third-party

> Unblinding" below.

Reviewer: ```suggestion
Unblinding" below).
```

Bob. Bob wants Victor to be able to see what and how much was sent. 

Victor, with Bob’s help, creates a watch-only wallet in Elements:

1. for each address of interest A, Victor runs the RPC `importaddress
<A>`

2. Bob exports the blinding key for address A using `dumpblindingkey
<A>` and shares it with Victor.

3. Victor imports the blinding key for the address using `importblindingkey <A> <blinding key>`

Once the blinding key is imported, the Elements wallet will treat the
Confidential address address as watch-only, and its outputs will be
visible in transaction details and in the wallet balance.  

Please note that if Bob reuses an address A, Victor will also be able
to see the amounts and values in any transaction sending to A. 

Alternatively, a watch-only wallet may import the master blinding key
of another wallet. The watch-only wallet would then be able to view
the UTXOs for any confidential address created by the original
wallet. 

Elements is not required to unblind a confidential transaction. Given
the confidential address’ blinding key, the rangeproof for the output
can be rewound, revealing the blinding factors and actual amounts and

> assets that were committed to.  

Reviewer: What are these two sentences really communicating? If you're not using Elements, how else would you do it? (other than with a pen and paper)

It sounds like you might really just be trying to say the following:
```
Anyone with the blinding key for an output's confidential address can rewind the range proof and reveal the blinding factors, along with the actual amounts and assets that were committed to.
```

Please see the [Elements Project
tutorial](https://elementsproject.org/elements-code-tutorial/advanced-examples)
for examples of how to unblind with Elements.

## Wallet Considerations

An Elements wallet has a "master blinding key", from which all
blinding keys for that wallet are deterministically derived. A
blinding key for an address is generated as HMAC_SHA256(master

> blinding key, <address ScriptPubKey>). See SLIP-0077 [6]. 

Reviewer: This is not rendered correctly (see the corresponding text in the rendered markdown document).

I believe you have to escape the angle brackets.

> blinding key for an address is generated as HMAC_SHA256(master
> blinding key, <address ScriptPubKey>). See SLIP-0077 [6]. 

Reviewer: This is not rendered correctly (see the corresponding text in the rendered markdown document).

I believe you have to escape the angle brackets.

Alternatively, perhaps this equation can go inside a code block, like this:

```suggestion
blinding key for an address is generated as `HMAC_SHA256(master
blinding key, <address ScriptPubKey>)`. See SLIP-0077 [6]. 
```

proof’s serialization in the transaction output.  A larger range
requires a larger proof. In Elements, the default size of the range
proof is 52 bits, allowing for a value of up to 2^52-1 satoshis to be
represented. In Elements, it is possible to create more than 21
million of an asset over multiple issuances, but it is not
possible to send more than 21 million in any one unconfidential output.  The range
proof parameters are not part of consensus, and may be overridden and
adjusted using the `ct_bits` elements configuration parameter. Reducing
the number of bits will reduce the size of a transaction, and  also
reduce the maximum provable value of any output. 

An Elements range proof is a Borromean ring signature [4] over
possible values of each digit in the base 4 representation of an
output value. Each digit requires the storage of 5 elliptic curve
points. Not including a fixed size header for the range proof, the
space required for a range proof in Elements is approximately 80 bytes
per bit of precision (default 52). 

The [secp256k1-zkp range proof implementation](https://github.com/BlockstreamResearch/secp256k1-zkp/blob/master/include/secp256k1_rangeproof.h)

> supports a maximum range of 64 bits. A range proof supporting a 64 bit

Reviewer: typo? 

```suggestion
supports a maximum range of 64 bits. A range proof supporting a 64 bit <number, range, etc.>
```

Author: thanks, fixed amended commit
